src/pip/_internal/req/constructors.py

```diff
@@ -368,6 +369,14 @@ def with_source(text):
             if add_msg:
                 msg += '\nHint: {}'.format(add_msg)
             raise InstallationError(msg)
+        else:
+            # Check that specifiers are PEP-440 compliant
+            for spec in req.specifier:
+                if not isinstance(spec, Specifier):
```

Comment:
What would the type of `spec` be, if not `Specifier`? Downstream distributors of pip perform “debundling” that changes the import path of `pip._vendor.packaging`, so we want to avoid `isinstance()` checks if possible, since they may return wrong results.

Reply:
spec is either a Specifier (if it parses as PEP-440 specifier) or a LegacySpecifier (if it does not). Based on Changelogs this has been part of the packaging API since 2014 (packaging 14.3) but probably my use of SpecifierSet iteration only came available in 2015 (15.2).

I'm happy to use another method as well if you don't trust results from packaging.